concordium-consensus/src/Concordium/Skov/MonadImplementations.hs

```diff
@@ -460,114 +475,131 @@ instance SkovTimerHandlers pv (SkovPassiveHandlers pv c m) c m where
+runSkovT comp =
+    runSkovTInternal $ runPersistentBlockStateMonad $ runPersistentTreeStateMonad $ runSkovT' comp
 
 evalSkovT :: (Monad m) => SkovT pv h c m a -> h -> SkovContext c -> SkovState c -> m a
-evalSkovT (SkovT a) h c = evalStateT (a h c)
+evalSkovT comp handler context sstate = fst <$> runSkovT comp handler context sstate
+
+askHandler :: (Monad m) => SkovT pv h c m h
+askHandler = SkovT $ lift $ lift $ SkovTInternal $ asks srHandler
 
 instance (Monad m) => MonadReader (SkovContext c) (SkovT pv h c m) where
-    ask = SkovT (\_ c -> return c)
-    local f (SkovT a) = SkovT (\h -> a h . f)
+    ask = SkovT $ lift ask
+    local f (SkovT (PersistentTreeStateMonad a)) = SkovT $ lift $ local f a
 
 instance MonadTrans (SkovT pv h c) where
-    lift a = SkovT (\_ _ -> lift a)
+    lift a = SkovT $ lift $ lift $ lift a
 
 instance (Monad m, SkovTimerHandlers pv h c m) => TimerMonad (SkovT pv h c m) where
     type Timer (SkovT pv h c m) = SkovHandlerTimer h
-    onTimeout timeout a = SkovT (\h _ -> lift (handleOnTimeout h timeout a))
-    cancelTimer t = SkovT (\h _ -> lift (handleCancelTimer h t))
+    onTimeout timeout a = do
+        h <- askHandler
+        lift (handleOnTimeout h timeout a)
+    cancelTimer t = do
+        h <- askHandler
+        lift (handleCancelTimer h t)
 
------------------------------------------------------------------------------
--- Inherit the instances from GlobalStateM
+instance (IsProtocolVersion pv) => MonadProtocolVersion (SkovT pv h c' m) where
+    type MPV (SkovT pv h c' m) = pv
 
-deriving via
-    SkovTGSM pv h c' m
-    instance
-        BlockStateTypes (SkovT pv h c' m)
+instance (c ~ SkovConfig pv gsconfig finconfig handlerconfig) => HasBlobStore (SkovContext c) where
+    blobStore = blobStore . scGSContext
+    blobLoadCallback = blobLoadCallback . scGSContext
+    blobStoreCallback = blobStoreCallback . scGSContext
 
-deriving via
-    SkovTGSM pv h c' m
-    instance
-        ( Monad m,
-          BlockStateQuery (SkovTGSM pv h c' m)
-        ) =>
-        BlockStateQuery (SkovT pv h c' m)
+instance (c ~ SkovConfig pv gsconfig finconfig handlerconfig, AccountVersionFor pv ~ av) => HasCache (AccountCache av) (SkovContext c) where
+    projectCache = projectCache . scGSContext
 
-deriving via
-    SkovTGSM pv h c' m
-    instance
-        ( Monad m,
-          AccountOperations (SkovTGSM pv h c' m)
-        ) =>
-        AccountOperations (SkovT pv h c' m)
+instance (c ~ SkovConfig pv gsconfig finconfig handlerconfig) => HasCache ModuleCache (SkovContext c) where
+    projectCache = projectCache . scGSContext
 
-deriving via
-    SkovTGSM pv h c' m
-    instance
-        ( Monad m,
-          ContractStateOperations (SkovTGSM pv h c' m)
-        ) =>
-        ContractStateOperations (SkovT pv h c' m)
+deriving instance
+    ( IsProtocolVersion pv,
+      MonadIO m,
+      c ~ SkovConfig pv gsconfig finconfig handlerconfig
+    ) =>
+    BlockStateQuery (SkovT pv h c m)
```

Comment:
These instances are not the best/cleanest since the magical gsconfig finconfig handlerconfig have to appear out of thin air.

Would it make sense to remove the `c` and directly parametrize `SkovT` by `finconfig` and `handlerconfig`? Or is that too noisy?

Reply:
I agree that it is anoying to have many parameters.

But I still think it is "magical" in the sense that the gsconfig, finconfig, handlerconfig magically appear and they are not part of the type, requiring further type system extensions to support.

I'm OK not changing this now.